src/main/java/org/jabref/gui/externalfiles/UnlinkedFilesDialogViewModel.java

```diff
@@ -0,0 +1,269 @@
+
+        treeRootProperty.setValue(null);
+    }
+
+    public void startSearch() {
+        Path directory = this.getSearchDirectory();
+        Filter<Path> selectedFileFilter = selectedExtension.getValue().dirFilter();
+
+        progressValueProperty.unbind();
+        progressTextProperty.unbind();
+
+        findUnlinkedFilesTask = new UnlinkedFilesCrawler(directory, selectedFileFilter, bibDatabase, preferences.getFilePreferences())
+                .onRunning(() -> {
+                    progressValueProperty.set(ProgressIndicator.INDETERMINATE_PROGRESS);
+                    progressTextProperty.setValue(Localization.lang("Searching file system..."));
+                    progressTextProperty.bind(findUnlinkedFilesTask.messageProperty());
+                    taskActiveProperty.setValue(true);
+                    treeRootProperty.setValue(null);
+                })
+                .onFinished(() -> {
+                    progressValueProperty.set(0);
+                    taskActiveProperty.setValue(false);
+                })
+                .onSuccess(root -> {
+                    progressValueProperty.set(0);
+                    taskActiveProperty.setValue(false);
+                    treeRootProperty.setValue(root);
+                });
+        findUnlinkedFilesTask.executeWith(taskExecutor);
+    }
+
+    public void startImport() {
+        List<Path> fileList = checkedFileListProperty.stream()
+                                                     .map(item -> item.getValue().getPath())
+                                                     .filter(path -> path.toFile().isFile())
+                                                     .collect(Collectors.toList());
+        if (fileList.isEmpty()) {
+            LOGGER.warn("There are no valid files checked");
+            return;
+        }
+        resultList.clear();
+
+        importFilesBackgroundTask = importHandler.importFilesInBackground(fileList)
+                                                 .onRunning(() -> {
+                                                     progressValueProperty.bind(importFilesBackgroundTask.workDonePercentageProperty());
+                                                     progressTextProperty.bind(importFilesBackgroundTask.messageProperty());
+                                                     taskActiveProperty.setValue(true);
+                                                 })
+                                                 .onFinished(() -> {
+                                                     progressValueProperty.unbind();
+                                                     progressTextProperty.unbind();
+                                                     taskActiveProperty.setValue(false);
+                                                 })
+                                                 .onSuccess(results -> {
+                                                     progressValueProperty.unbind();
```

Comment:
There is some code dup with the onFinished handler, is still really needed?

Reply:
onFinished is called when the task exited but with not sucesss

Reply:
If I understand https://github.com/JabRef/jabref/blob/c8f7be89ad704065474decb61851e42416caf88d/src/main/java/org/jabref/gui/util/BackgroundTask.java#L156-L162 correctly, then onFinished is always called regardless of if it was a success or an error.